### DISCUSSION TITLE
CPP compositing alpha onto solid background

### DISCUSSION THREAD

**Discussion by laserted007** [libvips/libvips#2280] [Q&A] [answered]
Greets - I'm trying to clamp the RGB of an existing png file with alpha, then finish it by overlaying the result onto a background of a different color. Dev environ is Debian 10 and CPP. (GCC 8.3)

What I've tried so far:
   ```
    VImage in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    VImage bgnd = VImage::black(300,300).invert();
    VImage clamp = (in > 250).ifthenelse(250,in);
```

the above is successful in clamping to a level of 250, and creating a separate background, in this case 255-white (inverted black).
The battle I'm having is in getting the background to stick to the original image.

I've tried flatten, composite, insert, another iteration of ifthenelse, and all of the outputs still retain the alpha and completely ignore the background.  Both the source, background and output are expected to be 300px square.

Recommendations?

Thanks!

**Comment by jcupitt** [comment]
Hi @laserted007,

I'm not sure I understand what you're trying to achieve. Could you post a test image and desired output?

**Reply by laserted007** [reply]
Thank you for the quick response, I am happy to elaborate.

We have available: "source_RGBA.png" It is a fixed known size, (ie 300x300) item with an  alpha channel (thus will include transparent regions) and RGB values spanning 0-255. Alpha isn't analog, it's only 0 or 255 based upon the source image.

The first step is to clamp the RGB to 250 max. Yes, I realize we are reducing the color gamut, but the slight loss is irrelevant to the requirement. This is an interim step that does not actually need to generate an output file, but I've included it as progress reference (clamped250_RGBA.png). At this stage, the alpha channel is still important to retain (at least I believe such) so we have a known difference between image content and transparent region. I am able to complete this step just fine with an ifthenelse method: `VImage clamp = (in > 250).ifthenelse(250,in); `

The next step is to convert the transparent region (any alpha wherein A<128 if you need numbers, keeping in mind the alpha only appears in this situation as either 0 or 255) to RGB=255,255,255 (white), while at the same time REMOVING the alpha channel. The equivalent is attached as "target_RGB.png"

![source_RGBA](https://user-images.githubusercontent.com/9323159/120651610-f9a79480-c44c-11eb-8ffd-863f6ca6fa9b.png)
(Above source_RGBA.png)

![clamped250_RGBA](https://user-images.githubusercontent.com/9323159/120651624-fd3b1b80-c44c-11eb-984a-a080a5fd6ba8.png)
(Above clamped250_RGBA.png, an interim reference)

![target_RGB](https://user-images.githubusercontent.com/9323159/120651634-ff9d7580-c44c-11eb-8a17-18f10cd1c27a.png)
(Above target_RGB.png)

You can tell a slight difference between the 250white (circles) in the target file and the 255white background. This is the intended effect.

My attempts thus far have typically bounced between results of a fully white artboard (no actual content, just 255white) or errors regarding image length (Tile mismatch) - which I believe is libvips trying to convert an alpha-inclusive and a non-alpha-inclusive VImage type, and failing at it.

This is actually my first day using libvips and I'm thrilled at how quick it was to pick up. (In the past I've been using opencv and ImageMagick, but those are not appropriate for this particular application, something fast and very lightweight is called for). I would love to see more sample code on the cpp binding usage (realizing that's in the works) as I'm definitely a try-it-to-learn kind of guy; I got some of the best help on the function/method use from the compiler errors, it would be helpful to translate that into a reference doc. (Unfortunately I find the many confused postings about various binding techniques complicates the understanding past useful knowledge.)

Guidance most appreciated!

TH.

**Comment by jcupitt** [comment]
Ah gotcha. libvips treats images as huge 3D arrays (width, height, bands) and doesn't really know anything about alpha. By convention, it's the last band, but it's just a convention. Not having special rules or special image types is supposed to make the behaviour more predictable.

When you clamp your image, you're clamping the alpha as well. I'd remove the alpha and process it separately. In Python:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])

# remove the alpha
alpha = image[3]
rgb = image[0:3]

# clamp rgb to 0 - 255
rgb = (rgb > 250).ifthenelse(250, rgb)

# replace transparent parts by 255
rgb = alpha.ifthenelse(rgb, 255)

rgb.write_to_file(sys.argv[2])
```

Run like this:

```
$ ./try329.py ~/pics/source_RGBA.png x.png
```

To make:

![x](https://user-images.githubusercontent.com/580843/120655582-bb28ce80-c47a-11eb-8b86-f5db0f0be2c4.png)

**Comment by jcupitt** [comment]
The C++ API docs have been fixed up for libvips 8.11 (due for release next week), which should help you. There's some sample output here:

https://libvips.github.io/libvips/API/8.11/cpp/index.html

**Reply by laserted007** [reply]
Thank you; I did happen upon the preview of 8.11 you noted above. My python, embarrassingly, is very very rusty. And apparently my Google-fu isn't doing too well today either - could you advise on the declaration of those array types (alpha and rgb) as cpp, kindly? They're not simply int arrays, are they? My read of ifthenelse.c infers it's a VipsImage, but that doesn't see right, either.

**Reply by jcupitt** [reply]
In pyvips, you can use `[]` to get bands from images. So `rgb[0]` will pull out the red band and return it as a one-band image. You can use array slicing to get sets of bands: `image[0:3]` means extract the first three bands (ie. RGB).

In C++ you can get single bands in the same way (vipscpp overloads `[]`), but there's no array slicing syntax, so you have to call `extract_bands` directly. Something like:

```C++
rgb = image.extract_bands(0, VImage::option()->set ("n", 3)); 
```

ie. extract three bands, starting with band 0.

**Comment by laserted007** [comment, marked as answer]
Awesome, thank you! I was getting (what I expect was) some alpha-compression-garbage using just the two part math you note above, which was showing up as a small "halo" or "trim-ring" around the perimeter of object(s), so I clamped the entire image first, then the alpha and rgb separately (a combo of our efforts, which works in all of my test cases, not sure if it's a globally-accepted solution.)  Below is my code snippet in case it assists anyone else:
```

  VImage in = VImage::new_from_file (argv[1], VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));     // load file
  VImage clamp = (in > 250).ifthenelse(250,in);                          // first clamping, entire RGBA. It helps get rid of the trim ring artifact.

  VImage alpha = clamp[3];                                                         // subarray the alpha "band" (element 4) into a standalone object
  VImage rgb = clamp.extract_band(0, VImage::option()->set ("n", 3));        // subarray the RGB bands (elements 1-3) into a standalone object

  alpha = (alpha < 128).ifthenelse(0,250);                 // clamp the alpha to binary of either 0 or 250, based upon an arbitrary middle of 127
  rgb = alpha.ifthenelse(rgb, 255);                           // flush the alpha through rgb, making a choice of either image (rgb) or 255white based upon alpha input
  rgb.write_to_file (argv[2]);                                        // output file (as png).

```
Thank you once again! As I play more with this, I'm liking it more and more. Still wrapping my head around bands versus channels, but that will come.

TH.

**Reply by jcupitt** [reply]
Bands and channels are the same thing. libvips is stuck with the "bands" name due to an accident of history.

**Reply by kleisauke** [reply]
https://github.com/libvips/libvips/wiki/vips-terminology may also be useful.
